sold.cc

```diff
@@ -1,985 +1,623 @@
+// RelocateSymbol_x86_64 rewrites r_offset of each relocation entries
+// because we decided locations of shared objects in DecideOffsets.
+void Sold::RelocateSymbol_x86_64(ELFBinary* bin, const Elf_Rel* rel, uintptr_t offset) {
+    const Elf_Sym* sym = &bin->symtab()[ELF_R_SYM(rel->r_info)];
+    auto [soname, version_name] = bin->GetVersion(ELF_R_SYM(rel->r_info), filename_to_soname_);
+
+    int type = ELF_R_TYPE(rel->r_info);
+    const uintptr_t addend = rel->r_addend;
+    Elf_Rel newrel = *rel;
+    if (bin->InTLS(rel->r_offset)) {
         const Elf_Phdr* tls = bin->tls();
         CHECK(tls);
-        CHECK(!tls_.data.empty());
-        auto found = tls_.bin_to_index.find(bin);
-        CHECK(found != tls_.bin_to_index.end());
-        const TLS::Data& entry = tls_.data[found->second];
-        if (off < tls->p_filesz) {
-            LOG(INFO) << "TLS data " << msg << " in " << bin->name() << " remapped " << HexString(off) << " => "
-                      << HexString(off + entry.file_offset);
-            off += entry.file_offset;
-        } else {
-            LOG(INFO) << "TLS bss " << msg << " in " << bin->name() << " remapped " << HexString(off) << " => "
-                      << HexString(off + entry.bss_offset);
-            off += entry.bss_offset;
-        }
-        return off;
-    }
-
-    void LoadDynSymtab(ELFBinary* bin, std::vector<Syminfo>& symtab) {
-        bin->ReadDynSymtab(filename_to_soname_);
-
-        uintptr_t offset = offsets_[bin];
-
-        for (const auto& p : bin->GetSymbolMap()) {
-            const std::string& name = p.name;
-            Elf_Sym* sym = p.sym;
-            if (IsTLS(*sym) && sym->st_shndx != SHN_UNDEF) {
-                sym->st_value = RemapTLS("symbol", bin, sym->st_value);
-            } else if (sym->st_value) {
-                sym->st_value += offset;
-            }
-            LOG(INFO) << "Symbol " << name << "@" << bin->name() << " " << sym->st_value;
-
-            Syminfo* found = NULL;
-            for (int i = 0; i < symtab.size(); i++) {
-                if (symtab[i].name == p.name && symtab[i].soname == p.soname && symtab[i].version == p.version) {
-                    found = &symtab[i];
-                    break;
-                }
-            }
-
-            if (found == NULL) {
-                symtab.push_back(p);
+        uintptr_t off = newrel.r_offset - tls->p_vaddr;
+        off = RemapTLS("reloc", bin, off);
+        newrel.r_offset = off + tls_offset_;
+    } else {
+        newrel.r_offset += offset;
+    }
+
+    LOG(INFO) << "Relocate " << bin->Str(sym->st_name) << " at " << rel->r_offset;
+
+    switch (type) {
+        // TODO(akawashiro) Check sym is invalid.
+        // This relocation type doesn't need any symbol therefore sym
+        // should be invalid, isn't it?
+        case R_X86_64_RELATIVE: {
+            newrel.r_addend += offset;
+            break;
+        }
+
+        // Even if we found the defined symbol in src_syms_, we cannot
```

Comment:
I like this comment, thanks! I think this applies to all relocation types. How about moving this to before `switch (type)`? I'd also slightly adjust phrasing or the comment

```
    // Even if we found a defined symbol in src_syms_, we cannot
    // erase the relocation entry. The address needs to be fixed at
    // runtime by ASLR function so we set RELATIVE to these resolved symbols.
    switch (type) {
```

Reply:
I see.